src/main/java/eu/interop/federationgateway/service/DiagnosisKeyBatchService.java

```diff
@@ -65,158 +61,140 @@
+      newBatchTag = formattedDate + "-" + (getSequenceNumberFromBatchTag(lastBatchEntity.get().getBatchName()) + 1);
 
-      log.error("Stop batching process, while try to batch {} keys, but the keys have different format versions",
-        diagnosisKeyEntitys.size());
-      return true;
+      lastBatchEntity.get().setBatchLink(newBatchTag);
+      diagnosisKeyBatchRepository.save(lastBatchEntity.get());
+    } else { // otherwise create the first batch for today
+      newBatchTag = formattedDate + "-1";
     }
-    return false;
+
+    newBatchEntity.setBatchName(newBatchTag);
+    return diagnosisKeyBatchRepository.save(newBatchEntity);
   }
 
-  private void updateDiagnosisKeys(List<DiagnosisKeyEntity> diagnosisKeyEntitys, DiagnosisKeyBatchEntity newBatch) {
-    diagnosisKeyEntitys.forEach(key -> key.setBatchTag(newBatch.getBatchName()));
-    diagnosisKeyEntityRepository.saveAll(diagnosisKeyEntitys);
+  /**
+   * Queries the database for unbatched uploaded keys until the maximum batch size is reached.
+   *
+   * @return a list of uploader batch tags that need to be put into one batch.
+   */
+  @Transactional(Transactional.TxType.MANDATORY)
+  public List<String> collectUploaderBatchTags() {
+    List<String> uploaderBatchTags = new ArrayList<>();
+    int newBatchSize = 0;
 
-    MDC.put("batchTag", newBatch.getBatchName());
-    MDC.put("diagnosisKeysCount", String.valueOf(diagnosisKeyEntitys.size()));
+    while (true) {
+      Optional<DiagnosisKeyEntity> unbatchedDiagnosisKey = uploaderBatchTags.isEmpty()
+        ? diagnosisKeyEntityRepository.findFirstByBatchTagIsNull()
+        : diagnosisKeyEntityRepository.findFirstByBatchTagIsNullAndUploaderBatchTagIsNotIn(uploaderBatchTags);
 
-    log.info("Batch created");
-  }
+      if (unbatchedDiagnosisKey.isEmpty()) {
+        // no more unprocessed keys
+        break;
+      }
+
+      String uploaderBatchTag = unbatchedDiagnosisKey.get().getUploader().getBatchTag();
+      int uploadBatchSize = diagnosisKeyEntityRepository.countAllByUploader_BatchTag(uploaderBatchTag);
 
-  private DiagnosisKeyBatchEntity saveBatches(Optional<DiagnosisKeyBatchEntity> lastEntry) {
-    ZonedDateTime currentDateTime = ZonedDateTime.now(ZoneOffset.UTC);
-    String formattedDate = currentDateTime.format(DateTimeFormatter.ofPattern("yyyyMMdd"));
-    var batchEntity = new DiagnosisKeyBatchEntity();
-
-    if (lastEntry.isPresent() && isBatchFromCurrentDay(lastEntry, currentDateTime)) {
-      batchEntity.setBatchName(formattedDate + "-" + getSequenceFromPreviousBatchAndCountOn(lastEntry));
-      lastEntry.get().setBatchLink(batchEntity.getBatchName());
-      // update the last batch by link to the new batch
-      diagnosisKeyBatchRepository.save(lastEntry.get());
-      log.info("successfully updated the linked batch entity");
-    } else {
-      batchEntity.setBatchName(formattedDate + "-1");
+      if (newBatchSize + uploadBatchSize <= properties.getBatching().getDoclimit()) {
```

Comment:
we should discuss about it again, the batch doc limit must always less than or equal to the upload limit
maybe we should take the same parameter